packages/xstate-react/src/useMachine.ts

```diff
@@ -0,0 +1,133 @@
+    Partial<UseMachineOptions<TContext, TEvent>> &
+    Partial<MachineOptions<TContext, TEvent>> = {}
+): [
+  State<TContext, TEvent, any, TTypestate>,
+  Interpreter<TContext, any, TEvent, TTypestate>['send'],
+  Interpreter<TContext, any, TEvent, TTypestate>
+] {
+  if (process.env.NODE_ENV !== 'production') {
+    const [initialMachine] = useState(machine);
+    if (machine !== initialMachine) {
+      console.warn(
+        'Machine given to `useMachine` has changed between renders. This is not supported and might lead to unexpected results.\n' +
+          'Please make sure that you pass the same Machine as argument each time.'
+      );
+    }
+  }
+
+  const {
+    context,
+    guards,
+    actions,
+    activities,
+    services,
+    delays,
+    state: rehydratedState,
+    ...interpreterOptions
+  } = options;
+
+  // Capture all actions (side-effects) to be executed.
+  // These will be flushed when they are executed, and avoids the issue of batched events
+  // sent to the interpreter, which might ignore actions.
+  const actionStatesRef = useRef<Array<State<TContext, TEvent>>>([]);
+
+  // Keep a single reference to the invoked machine (the service)
+  const service = useConstant(() => {
+    const machineConfig = {
+      context,
+      guards,
+      actions,
+      activities,
+      services,
+      delays
+    };
+    const createdMachine = machine.withConfig(machineConfig, {
+      ...machine.context,
+      ...context
+    } as TContext);
+
+    // Ensure that actions are not executed (until useEffect() below)
+    interpreterOptions.execute = false;
+
+    return interpret(createdMachine, interpreterOptions).start(
+      rehydratedState ? State.create(rehydratedState) : undefined
+    );
+  });
+
+  // Make sure actions and services are kept updated when they change.
+  // This mutation assignment is safe because the service instance is only used
+  // in one place -- this hook's caller.
+  useEffect(() => {
+    Object.assign(service.machine.options.actions, actions);
+  }, [actions]);
+  useEffect(() => {
+    Object.assign(service.machine.options.services, services);
+  }, [services]);
+
+  // Initialize the state with the initial state.
+  const [state, setState] = useState(service.state);
+
+  useEffect(() => {
+    // Whenever a new state is emitted from the service,
+    // update the state with that state, but only if
+    // that state has changed.
+    service.onTransition((currentState) => {
+      if (currentState.changed) {
+        // capture side-effects to be executed
+        actionStatesRef.current.push(currentState);
+
+        // change state
+        setState(currentState);
+      }
+    });
+
+    return () => {
+      service.stop();
+    };
+  }, []);
+
+  useEffect(() => {
+    // Flush all actions to be executed (per state)
+    actionStatesRef.current.forEach((actionState) => {
+      // Execute all actions for the captured state
+      service.execute(actionState);
```

Comment:
an additional quirk of this approach which I have thought of right now:
1. you send an event to itself, which is supposed to transition you from A to B - but it gets postponed until useEffect
2. other transitions happen in between, taking you from A to C
3. the postponed send gets executed and sends the event to a machine in a completely different state that it was supposed to, this can lead to no transition being taken or a completely different transition being taken

Reply:
That's fine; `send(...)` is not supposed to be zero-time. You should always anticipate that an event can be sent before `send(...)` is flushed.

For immediate deterministic behavior, that's what `raise(...)` is for.